I've fixed vulns as a developer!

FedeLu01 opened the pull request.

theowni commented:
Hi @FedeLu01,
great job!
thank you for the submission.
I have the following feedback for you. When you adjust your code, I will happily accept your submission!

**Minor Recommendations**:
1. In "Level 2", you could directly use `current_user.username` for obtaining a user's object. But you would need to remove `username` field from `UserUpdate` model already used. It would be simpler in implementation and prevent issues in future.

**Required Fixes**
1. In "Level 4" you check if the `image_url` contains the `https://bucket.s3.amazonaws.com`. Actually, this check could be bypassed with the following URL:
 - `https://bucket.s3.amazonaws.com.malicious.com`

As a result, an attacker could host some malicious content using own domain as the above domain will also pass your check.
I'd suggest to use a library such as `urllib.parse` to extract network location from the URL and then validate if it's equal to allowed locations.
 
btw.  you could also validate the extension of the file, not only content-type.

2. In "Level 5" in your fixed code I don't see where you validate that `parameters` received from request are in `allowed_parameters`.

FedeLu01 commented:
Hi @theowni!! 
Thanks a lot for your feedback! I've used it to improve my solutions, the changes are already implemented. 
Can you please check?

theowni commented:
"Level 4" looks good enough for the purposes of this challenge.
In terms of "Level 5", I have a suggestion for you combining your current and previous approach. In your previous approach you had a list of allowed parameters. It would be recommended to keep it as you can validate what parameters can be passed to `df -h` command. Also, you still use `shell=True` flag, so I'd remove it as it's not needed.

Here is my suggestion combining your both approaches:

```
def get_disk_usage(parameters: str):
    allowed_parameters = [
        "-a", "--all", "-B", "--block-size", "-h", "--human-readable", 
        "-H", "--si", "-i", "--inodes", "-k", "-l", "--local", 
        "--no-sync", "--output", "-P", "--portability", "--sync", 
        "--total", "-t", "--type", "-T", "-x", "--exclude-type"
    ]
    
    # split parameters string to a list of params and validate them against allowed_parameters list
    input_params = parameters.split()
    for param in input_params:
        if param not in allowed_parameters:
            raise ValueError(f"Invalid parameter: {parameters}")
        
    command = ["df", "-h"]
    command.extend(input_params)   # extend command list with validated parameters

    try:
        result = subprocess.run(
            command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True  # here, I removed shell=True flag
        )
        usage = result.stdout.strip().decode()
    except subprocess.CalledProcessError as e:
        raise Exception(f"Command failed: {e.stderr.decode()}")
    except Exception as e:
        raise Exception(f"An unexpected error occurred: {e}")

    return usage
```

FedeLu01 commented:
Your suggestions makes total sense, thank you! 
Level 5: I did try to validate the parameters against available parameters in df command, but I didn't get to think in using .split method to grab the user input, that's why I've deleted it from my code in the first iteration.
Level 4: Aside from validating the hostname and port as I did in this case, what could be a production implementation? Is it a good idea to validate the scheme as well?
I've applied the fixes, please check them! :)

theowni commented:
yes, in terms of URL validation I always recommend the following checks:
* validate the domain (or IP address depending on the use case)
* validate the port
* validate the path if the context requires that. In our case we can skip this.

In your code you already validate the domain and the port, so your solution is satisfying :)

Could you make one last change in your PR to include only changes in "HALL_OF_FAME.md" file? Currently, this PR also includes modified `.py` files.

FedeLu01 commented:
Thanks for all your feedback @theowni!
This was a really fun challenge! I've learned a lot :)

theowni commented:
Awesome! I'm happy to hear that.

Let me know when you adjust this PR, so I will be able to merge your request with modified HALL_OF_FAME.md.

FedeLu01 commented:
Ok, done!
Thanks again! 👍

theowni commented:
I've just noticed that your link is not working 😅
https://github.com/FedeLu01/Damn-Vulnerable-RESTaurant-API-Game/blob/main/vulnerability-fixes.md

FedeLu01 commented:
I was thinking, can you merge so I then add the file to the forked repo?
Or can you explain any alternative to do this? 
Thanks in advance!

theowni commented:
ok, I merged the PR. You're in the Hall of Fame now! Congrats!
Let me know when you add your writeup to the fork and I will close this PR.

FedeLu01 commented:
Done! 
Again, thanks a lot for this challenge! It was really fun and an amazing learning experience.

theowni commented:
thank you!
I appreciate that feedback very much!
